File: src/main/java/org/hildan/bots/riseoflords/RolAutomizer.java
package org.hildan.bots.riseoflords;

import java.io.Console;
import java.io.IOException;
import java.time.Duration;

import org.hildan.bots.riseoflords.config.Config;
import org.hildan.bots.riseoflords.config.Config.BadConfigException;
import org.hildan.bots.riseoflords.sequencing.LoginException;
import org.hildan.bots.riseoflords.sequencing.Sequence;
import org.hildan.bots.riseoflords.util.Log;

public class RolAutomizer {

    private static final String TAG = RolAutomizer.class.getSimpleName();

    public static void main(String[] args) {
        try {
            launch(args);
        } catch (LoginException e) {
            Log.get().e(TAG, "\nLogin failed for user ", e.getUsername());
        } catch (Exception e) {
            Log.get().e(TAG, "\nUNCAUGHT EXCEPTION: ", e.getMessage());
            e.printStackTrace(System.err);
        }
        waitForEnter(null);
    }

    public static void launch(String[] args) throws LoginException {
        if (args.length == 0) {
            System.out.println("No config file provided: you must provide a .rol file to open.");
            System.out.println();
            System.out.println("More info at https://github.com/joffrey-bion/rol-automizer");
            return;
        }
        final String filename = args[0];

        Config config;
        try {
            config = Config.loadFromFile(filename);
        } catch (final BadConfigException e) {
            Log.get().e(TAG, "Error reading config file: ", e.getMessage());
            return;
        } catch (final IOException e) {
            try {
                config = Config.loadFromResource(filename);
            } catch (IOException | BadConfigException e2) {
                Log.get().e(TAG, "Error reading config file: ", e.getMessage());
                return;
            }
        }
        Log.get().title(TAG, "CONFIG");
        Log.get().i(TAG, config);
        System.out.println();

        final Sequence sequence = new Sequence(config);
        for (int i = 0; config.unlimitedAttacks() || i < config.getNbOfAttacks(); i++) {
            Log.get().title(TAG, String.format("ATTACK SESSION %d/%d", i + 1, config.getNbOfAttacks()));
            sequence.start();
            if (config.unlimitedAttacks() || i + 1 < config.getNbOfAttacks()) {
                // more attacks are waiting
                System.out.println();
                waitForNextAttack(config.getTimeBetweenAttacks());
            }
        }
        System.out.println();
        Log.get().i(TAG, "End of attacks.");
    }

    public static void waitForEnter(String message, Object... args) {
        final Console c = System.console();
        if (c != null) {
            if (message != null) {
                c.format(message, args);
            } else {
                c.format("\nPress ENTER to exit.\n");
            }
            c.readLine();
        }
    }

    private static void waitForNextAttack(Duration duration) {
        try {
            long millis = duration.toMillis() % 1000;
            Thread.sleep(millis);
            duration = duration.minusMillis(millis);

            while (!duration.isZero() && !duration.isNegative()) {
                Thread.sleep(1000);
                duration = duration.minusSeconds(1);
                printDuration(duration);
            }
            System.out.print("\r");
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    private static void printDuration(Duration d) {
        final long hours = d.toHours();
        final long minutes = d.minusHours(hours).toMinutes();
        final long seconds = d.minusHours(hours).minusMinutes(minutes).toMillis() / 1000;
        System.out.print("\r");
        System.out.print(String.format("   Next attack session in %s%02d:%02d...", (hours > 0 ? hours + ":" : ""),
                minutes, seconds));
    }
}

File: src/main/kotlin/org/hildan/bots/riseoflords/RolAutomizer.kt
package org.hildan.bots.riseoflords

import com.github.ajalt.clikt.core.*
import com.github.ajalt.clikt.output.*
import com.github.ajalt.clikt.parameters.options.*
import com.github.ajalt.clikt.parameters.types.*
import org.hildan.bots.riseoflords.client.*
import org.hildan.bots.riseoflords.config.*
import org.hildan.bots.riseoflords.sequencing.*
import org.slf4j.*
import kotlin.time.*
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.seconds

object RolAutomizer

fun main(args: Array<String>) = RolAutomizerCommand().main(args)

private val logger = LoggerFactory.getLogger(RolAutomizer::class.java)

class RolAutomizerCommand : CliktCommand() {

    init {
        context {
            helpFormatter = { MordantHelpFormatter(it, showDefaultValues = true) }
        }
    }

    private val username by option(
        "-u",
        "--username",
        help = "your Rise of Lords username",
        envvar = "ROL_USERNAME",
    ).required()

    private val password by option(
        "-p",
        "--password",
        help = "your Rise of Lords password",
        envvar = "ROL_PASSWORD",
    ).prompt("Enter your password:", hideInput = true)

    private val minRank by option(
        "-m",
        "--min-rank",
        help = "the minimum rank of the players to attack",
    ).int().default(400)

    private val maxRank by option(
        "-M",
        "--max-rank",
        help = "the maximum rank of the players to attack",
    ).int().default(2_200)

    private val minGold by option(
        "-g",
        "--min-gold",
        help = "the minimum gold of the enemy player to consider an attack worth it",
    ).long().default(400_000L)

    private val maxTurns by option(
        "-t",
        "--max-turns",
        help = "the maximum number of turns to use during an attack session",
    ).int().default(40)

    private val repairPeriod by option(
        "-r",
        "--repair-period",
        help = "the number of attacks between weapon repairs",
    ).int().default(5, "every 5 attacks")

    private val storageThreshold by option(
        "--storage-threshold",
        help = "the threshold above which we need to store the current gold into the chest",
    ).int().default(300_000).check("cannot store less than 300k into the chest") { it >= 300_000 }

    private val nbOfAttacks by option(
        "--attacks-count",
        help = "the number of attack sessions to perform",
    ).int().default(1)

    private val timeBetweenAttacks by option(
        "--rest-time",
        help = "the number of hours to wait between attack sessions",
        metavar = "HOURS",
    ).int().convert { it.toLong().hours }.default(12.hours)

    override fun run() {
        val config = Config(
            account = Account(username, password),
            playerFilter = PlayerFilter(minRank, maxRank, minGold),
            attackParams = AttackParams(maxTurns, repairPeriod, storageThreshold),
            nbOfAttacks = nbOfAttacks,
            timeBetweenAttacks = timeBetweenAttacks,
        )
        val attacks = AttackManager(config)
        repeat(nbOfAttacks) {
            logger.info("Starting attack session {}/{}", it + 1, nbOfAttacks)
            try {
                attacks.startAttackSession()
            } catch (e: LoginException) {
                logger.error("Login failed for user {}", e.username)
            } catch (e: Exception) {
                logger.error("UNCAUGHT EXCEPTION", e)
            }
            if (it + 1 < nbOfAttacks) {
                // more attacks are waiting
                waitForNextAttack(timeBetweenAttacks)
            }
        }
        logger.info("End of attacks")
    }
}

private fun waitForNextAttack(duration: Duration) {
    var remaining = duration
    try {
        printDuration(remaining)
        repeat(duration.inWholeSeconds.toInt()) {
            Thread.sleep(1000)
            remaining -= 1.seconds
            printDuration(remaining)
        }
        print("\r")
    } catch (e: InterruptedException) {
        Thread.currentThread().interrupt()
    }
}

private fun printDuration(d: Duration) {
    d.toComponents { h, m, s, _ ->
        print("\r")
        print(String.format("   Next attack session in %s%02d:%02d...", if (h > 0) "$h:" else "", m, s))
    }
}
